File: 백준/Gold/1987. 알파벳/알파벳.py
import sys


sin = sys.stdin.readline
R, C = map(int, sin().split())
alphabets = [list(sin().rstrip()) for _ in range(R)]
max_range, cur_range = 1, 1

"""
이미 방문한 위치라도, 다른 탐색에서는 자신의 알파벳 리스트에 포함되어 있지 않으면 방문할 수 있음
그리고 한 탐색 시도에서 방문했던 점은 시작점 포함해서 방문했던 점은 방문기록 리스트에 포함되어서 중복 탐색이 자체적으로 예방 가능함

그리고 bfs를 통해 매 탐색마다 리스트를 저장하면 시간 초과가 발행하므로, dfs를 사용하기로 함..
따라서 bfs에서 visited 사용하지 않던 것과 다르게 사용함
"""

# 상, 하, 좌, 우
mx = [1, -1, 0, 0]
my = [0, 0, -1, 1]
visited = [False] * (ord('Z') - ord('A') + 1)

def dfs(x, y):
    start = alphabets[y][x]
    visited[ord(start) - ord('A')] = True
    global cur_range
    global max_range

    for m in range(4):
        nx, ny = x + mx[m], y + my[m]

        if 0 <= nx < C and 0 <= ny < R:
            next_alphabet = ord(alphabets[ny][nx]) - ord('A')

            if visited[next_alphabet]:
                continue

            cur_range += 1
            max_range = max(cur_range, max_range)

            dfs(nx, ny)


    # 범위에 각별히 조심하기, 내 주변에서 4방향 탐색을 돌 동안에는 거리나 방문기록 정보를 가지고 있어야 함
    cur_range -= 1
    visited[ord(start) - ord('A')] = False


dfs(0, 0)
print(max_range)

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;
import java.util.stream.Stream;

public class Main {
	static int row;	// 행
	static int col;	// 렬
	static int[][] arr;	// 알파벳 지도
	static int[] dy = {-1, 1, 0, 0};	// 각각 상 하 담당
	static int[] dx = {0, 0, -1, 1};	// 각각 좌 우 담당
	static boolean[] visited;
	static int step = 1, max = 1;	// 임시로 세는 용도의 걸음 수와, 답으로 쓸 최대 걸음 수

	public static void main(String[] args) throws IOException {
		String[] s;
		String str;
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
		
		s = br.readLine().split(" ");
		row = Integer.parseInt(s[0]);
		col = Integer.parseInt(s[1]);
		arr = new int[row][col];
		
		for(int i = 0; i < row; i++) {
			str = br.readLine();
			for(int j = 0; j < col; j++) {
				arr[i][j] = (int)str.charAt(j) - 'A';	// 정수 배열에 char 저장하기 위해 A를 빼면 알파벳 A가 0으로 저장됨
			}
		}
		visited = new boolean[26];
		dfs(0, 0);
		bw.write(max+"");
		bw.flush();
	}
	
	public static void dfs(int y, int x) {
		int start = arr[y][x];
		visited[start] = true;
		
		for(int i = 0; i < 4; i++) {
			int next_y = dy[i] + y;
			int next_x = dx[i] + x;
			
			if(next_x < 0 || next_y < 0 || next_x >= col || next_y >= row) {
				continue;
			}
			
			int next = arr[next_y][next_x];
			if(visited[next])
				continue;
			
			max = Math.max(max, ++step);
			
			dfs(next_y, next_x);
		}
		// 다른 경로에서의 접근을 위해서 초기화하기
		step--;
		visited[start] = false;
	}
}
